cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java

```diff
@@ -336,6 +310,73 @@ public CommandLineCompleter getValueCompleter() {
+        final CommandLineCompleterFactory factory = globalOpProps == null ? null : globalOpProps.get(prop.getName());
+        if (factory != null) {
+            final String propName = ctx.getParsedCommandLine().getPropertyValue(Util.NAME);
+            if (Util.WRITE_ATTRIBUTE.equals(operationName) && propName != null) {
+                return getFilenameCompleter(ctx, address, propName);
+            }
+            if (propCompleter == null) {
+                propCompleter = factory.createCompleter(address);
+            }
+        }
+        if (propCompleter == null) {
+            final ModelNode typeNode = prop.getValue().get(Util.TYPE);
+            if (typeNode.isDefined() && ModelType.BOOLEAN.equals(typeNode.asType())) {
+                return SimpleTabCompleter.BOOLEAN;
+            }
+            if (prop.getValue().get(Util.FILESYSTEM_PATH).asBoolean()) {
+                return Util.isWindows() ? new WindowsFilenameTabCompleter(ctx) : new DefaultFilenameTabCompleter(ctx);
+            }
+            if (prop.getValue().has(Util.VALUE_TYPE)) {
+                final ModelNode valueTypeNode = prop.getValue().get(Util.VALUE_TYPE);
+                try {
+                    // the logic is: if value-type is set to a specific type
+                    // (i.e. doesn't describe a custom structure)
+                    // then if allowed is specified, use it.
+                    // it might be broken but so far this is not looking clear to me
+                    valueTypeNode.asType();
+                    if (prop.getValue().has(Util.ALLOWED)) {
+                        return getAllowedCompleter(prop);
+                    }
+                } catch (IllegalArgumentException e) {
+                    // TODO this means value-type describes a custom structure
+                    return new ValueTypeCompleter(prop.getValue());
+                }
+            }
+            if (prop.getValue().has(Util.ALLOWED)) {
+                return getAllowedCompleter(prop);
+            }
+        }
+        return propCompleter;
+    }
+
+    private CommandLineCompleter getFilenameCompleter(CommandContext ctx, OperationRequestAddress address, String propName) {
+        final ModelNode req = new ModelNode();
+        final ModelNode addrNode = req.get(Util.ADDRESS);
+        for (OperationRequestAddress.Node node : address) {
+            addrNode.add(node.getType(), node.getName());
+        }
+        req.get(Util.OPERATION).set(Util.READ_RESOURCE_DESCRIPTION);
+        try {
+            ModelNode response = ctx.getModelControllerClient().execute(req);
+            final ModelNode result = response.get(Util.RESULT);
+            if (result.isDefined()) {
+                ModelNode allAttrs = result.get(Util.ATTRIBUTES);
+                if (allAttrs.isDefined()) {
+                    final ModelNode propDescr = allAttrs.get(propName);
+                    if (propDescr.isDefined() && propDescr.get(Util.FILESYSTEM_PATH).asBoolean()) {
```

Comment:
If you want, L365 - 370 can be simplified down to

if (response.get(Util.RESULT, Util.ATTRIBUTES, propName, Util.FILESYSTEM_OATH).asBoolean()) {